### PR TITLE
DEV: extract channel. joined_by method

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
This only moves code around and doesn't change any behavior. This does two things:

1. Extracts the `channel.joined_by?` methods
2. Uses term "members" instead of "participants" for chat members

### CODE REVIEW

File: plugins/chat/app/models/chat/channel.rb [hunk 1]
```diff
@@ -107,6 +107,19 @@ def self.ensure_consistency!
       update_counts
     end
 
+    def can_be_joined_by?(user)
+      guardian = Guardian.new(user)
```
Comment (jjaffeux) [inline plugins/chat/app/models/chat/channel.rb:111]: User is supposed to have a guardian: `user.guardian` (which is a shorthand for what you do here)

File: plugins/chat/app/models/chat/channel.rb [hunk 2]
```diff
@@ -107,6 +107,19 @@ def self.ensure_consistency!
       update_counts
     end
 
+    def can_be_joined_by?(user)
```
Comment (jjaffeux) [inline plugins/chat/app/models/chat/channel.rb:110]: shouldn't this be a guardian?

Comment (AndrewPrigorshnev) [reply plugins/chat/app/models/chat/channel.rb:110]: It probably should, I think it should be something like `user.can_join_chat_channel?` but the thing is we already have such method, we call it here in implementation:

    def can_be_joined_by?(user)
      guardian = Guardian.new(user)
      guardian.can_chat? && guardian.can_join_chat_channel?(self)
    end

I think we should refactor the guardian like this: https://github.com/discourse/discourse/pull/21812.

Someone who cannot chat is also not able to join chat channels.

File: plugins/chat/app/models/chat/channel.rb [hunk 3]
```diff
@@ -107,6 +107,18 @@ def self.ensure_consistency!
...
+      user.guardian.can_chat? && user.guardian.can_join_chat_channel?(self)
+    end
+
+    def joined_by?(user)
+      user.user_chat_channel_memberships.any? do |membership|
```
Comment (jjaffeux) [inline plugins/chat/app/models/chat/channel.rb:115]: maybe we add a `strict_loading` here?

it has huge potential for N+1 otherwise

**Comment by jjaffeux** [comment]
We should try to resume it soon so it doesn’t stale too much please

File: plugins/chat/app/models/chat/channel.rb (hunk 3 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/app/models/chat/channel.rb]: Makes sense 👍.

File: plugins/chat/app/models/chat/channel.rb (hunk 2 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/app/models/chat/channel.rb]: https://github.com/discourse/discourse/pull/21812 is merged now, so we don't need this method anymore and I've removed it.

We can use `guardian.can_join_chat_channel?` when needed instead.

File: plugins/chat/app/models/chat/channel.rb [hunk 4]
```diff
@@ -119,6 +119,14 @@ def self.ensure_consistency!
       update_user_counts
     end
 
+    def joined_by?(user)
+      user.user_chat_channel_memberships.strict_loading.any? do |membership|
+        predicate = membership.chat_channel_id == id
+        predicate = predicate && membership.following == true if public_channel?
```
Comment (jjaffeux) [inline plugins/chat/app/models/chat/channel.rb:125]:
```suggestion
        predicate = predicate && membership.following if public_channel?
```

**Review by jjaffeux — APPROVED** [review]

File: plugins/chat/app/models/chat/channel.rb (hunk 4 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/app/models/chat/channel.rb]: fixed

**Pull request merged**